src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs

```diff
@@ -856,7 +869,13 @@ private void ThrowValidationException()
         static internal object UnsafeGetUninitializedObject(Type type)
         {
 #if !NET_NATIVE
-            return TryGetUninitializedObjectWithFormatterServices(type) ?? Activator.CreateInstance(type);
+            bool hasDefaultConstructor;
+            if (!_typeHasDefaultConstructorMap.TryGetValue(type, out hasDefaultConstructor))
+            {
+                hasDefaultConstructor = type.GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, Array.Empty<Type>()) != null;
+                _typeHasDefaultConstructorMap[type] = hasDefaultConstructor;
+            }
+            return hasDefaultConstructor ? Activator.CreateInstance(type) : TryGetUninitializedObjectWithFormatterServices(type);
```

Comment:
You're right. We should keep the same behavior. Although I recall that the previous condition is mainly for safety but I cannot confirm that.